ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+    /// when `new_offset` is larger that the current `offset`, the function does nothing
+    fn reset_offset(&mut self, new_offset: RecordId, segment_length: usize) {
+        if new_offset < self.offset {
+            let mut new_offset = usize::from(new_offset);
+            // if segment_length is less than 256, redefine new offset such that we don't need to rearrange the existing segments
+            if 256 % segment_length == 0 {
+                // distance between new_offset and old offset is rounded up to a multiple of 256
+                new_offset = ((usize::from(self.offset) - new_offset + 255) >> 8) << 8;
+            }
+            let extension_length = ((usize::from(self.offset) - new_offset) >> 8) * segment_length;
+            // use existing vec and add enough space in front
+            let vec_previous = &mut self.vec;
+            let mut vec: Vec<Option<UnverifiedValues>> =
+                Vec::with_capacity(vec_previous.len() + extension_length);
+            vec.resize(extension_length, None);
+            vec.append(vec_previous);
+            self.offset = RecordId::from(new_offset);
+        }
+    }
+
+    /// `extend_vec` allows to extend a store to be able to hold a larger `RecordId`
+    fn extend_vec(&mut self, extension_length: usize) {
+        let vec_previous = &mut self.vec;
+        let new_length = vec_previous.len() + extension_length;
+        let mut vec: Vec<Option<UnverifiedValues>> = Vec::with_capacity(new_length);
+        vec.append(vec_previous);
+        vec.resize(new_length, None);
+    }
+
+    /// `insert_segment` allows to include a new segment in `UnverifiedValuesStore`
+    ///
+    /// ## Panics
+    /// Panics when segments have different lengths across records
+    fn insert_segment(&mut self, record_id: RecordId, segment: Segment) {
+        // initialize `segment_size` when necessary
+        if self.segment_size.is_none() {
+            self.initialize_segment_size(segment.len());
+        }
+        // check segment size
+        debug_assert_eq!(segment.len(), self.segment_size.unwrap());
+        // allocate if vec is not allocated yet
+        if self.vec.is_empty() {
+            self.allocate_vec();
+        }
+        // check offset
+        if record_id < self.offset {
+            // recover from wrong offset, expensive
+            self.reset_offset(record_id, segment.len());
+        }
+        let position_raw = usize::from(record_id) - usize::from(self.offset);
+        let length = segment.len();
+        let position_vec = (length * position_raw) >> 8;
+        // check size of store
+        if self.vec.len() < position_vec + (length >> 8) {
+            // recover from wrong size
+            // increase size of store to twice the size + position of end of the segment to be included
+            // expensive, ideally use initialize with correct length
+            self.extend_vec(2 * self.vec.len() - (position_vec + (length >> 8)));
```

Comment:
I think it would be better to just do `vec.resize(position_vec + (length >> 8))` unless/until we have evidence that the built-in allocation handling is not doing something reasonable.